codalab/lib/beam/blobstorageuploader.py

```diff
@@ -48,12 +51,13 @@ def put(self, data):
   def _write_to_blob(self, data):
     # block_id's have to be base-64 strings normalized to have the same length.
     block_id = base64.b64encode('{0:-32d}'.format(self.block_number).encode()).decode()
-    
-    self._blob_to_upload.stage_block(block_id, data)
+    # put the blob content to server, blob is uncommitted
+    self.all_tasks.append(self.thread_pool.submit(self._blob_to_upload.stage_block, block_id, data))
     self.block_list.append(BlobBlock(block_id))
     self.block_number = self.block_number + 1
 
   def finish(self):
     # The buffer will have a size smaller than MIN_WRITE_SIZE, so its contents can fit into memory.
     self._write_to_blob(self.buffer.read())
+    wait(self.all_tasks, return_when=ALL_COMPLETED)
```

Comment:
You should call `self.executor.shutdown()` here -- see https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown